tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/BuilderTest.scala

```diff
@@ -31,6 +31,157 @@ import scala.collection.immutable.SortedMap
+ * where `c /= a`. Note that the following is _not_ ill-typed:
+ * {{{
+ *    (
+ *        builder.name("x", a),
+ *        builder.name("y", FunT1(a, InvalidTypeMethods.differentFrom(b))),
+ *    )
+ * }}}
+ * since `b` isn't checked against any other argument, unlike `a`, if we swap `b` with some `c` we get arguments with
+ * types `a` and `a -> c`, which is still valid w.r.t. the signature of `foo`, and would return some value of type `c`.
+ * Some operators, for example those that take one argument of any type, have no ill-typed inputs. In those cases,
+ * {{{
+ *    def mkIllTyped(@unused tparam: TParam): Seq[T] = Seq.empty
+ * }}}
+ * </li>
+ *
+ * <li> A validator test `resultIsExpected`. It should have the type `TParam => ResultT => Boolean`, i.e., if `method`
+ * produces a value of type [[TBuilderInternalState]][ResultT] (recall, [[TBuilderInstruction]] is an alias for
+ * [[TBuilderInternalState]][TlaEx]), then `resultIsExpected` is a predicate over values of type `ResultT`
+ * (parameterized by `TParam`), that should return true iff the value produced by `method` meets expectations. In the
+ * vast majority of cases (excluding e.g. declaration constructors), `ResultT == TlaEx`. You can write your own, but the
+ * standard way is to use [[expectEqTyped]], which tests whether the builder constructs an [[OperEx]] expression, with
+ * the correct operator, arguments, and type tag. </li>
+ *
+ * </ol>
+ *
+ * To use [[expectEqTyped]], four things are needed:
+ *
+ * <ol>
+ *
+ * <li> The expected operator `op`. In our case, `foo`. </li>
+ *
+ * <li> A parametric constructor for the expected arguments. You should always use `mkWellTyped`. </li>
+ *
+ * <li> A transformer to turn a tuple of arguments into a sequence of arguments `toSeq`. This is required, due to the
+ * underlying shapeless framework, which allows us to generalize tests over all method arities. In the vast majority of
+ * cases, you can use the method in [[BuilderTest.ToSeq]], corresponding to the method arity. In our case,
+ * [[BuilderTest.ToSeq.binary]]. </li>
+ *
+ * <li> A predictor of the type of the return value `resType`, parameterized by the type parameters. In our case {
+ * {{{
+ * case (_, b) => b}
+ * }}}
+ * </li>
+ *
+ * </ol>
+ *
+ * In summary, the test for `builder.foo` is:
+ * {{{
+ * def resultIsExpected = expectEqTyped[TParam, T](
+ *   foo,
+ *   mkWellTyped,
+ *   ToSeq.binary,
+ *   { case (_, b) => b },
+ * )
+ * }}}
+ *
+ * Lastly, we need to define a generator for `TParam`. In most cases, you can use one of the generators defined in
+ * [[BuilderTest.Generators]]. In our case [[BuilderTest.Generators.doubleTypeGen]] for `(TlaType1, TlaType1)`.
+ *
+ * Our test then looks as follows:
+ * {{{
+ * checkRun(Generators.doubleTypeGen)(
+ *   runBinary(
+ *     builder.foo,
+ *     mkWellTyped,
+ *     mkIllTyped,
+ *     resultIsExpected,
+ *   )
+ * )
+ * }}}
```

Comment:
This does not look like a property based test to me, in particular because I see no sign of a property that we are testing for. Should the `resultIsExpected` be thought of as specifying the property that should hold? If so, it would probably help to use that language throughout the docs here (and in the implementation). 

In the property based tests I am familiar with, we find a generator, then a test that takes an arbitrary value coming from the generator that asserts some property that holds over an expression applied to the result of the generator. I *think* we kind of have the shape of that here, but its no clear to me that we actually do or whether this is the right way to think of it. If it *is* the right way to think of this, then I'm not sure why the names of the operators and behaviors don't reuse the established conventions of the PBT approach.

Reply:
The properties are outlined here:
https://github.com/informalsystems/apalache/blob/9346302d6c89fcc7ea13667a23fdc80c74795957/tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/BuilderTest.scala#L30-L33

Reply:
Or more loosely, the property is "builder builds type/scope-correct expressions, and only type/scope-correct expressions"

Reply:
If this helper is only for testing one property, then I'd change the intro to this package to explicitly state this limited scope. It's sounds like it's a utility for writing general purpose property based tests of built expressions, but iiuc from your reply, it just tests that one particular, fixed set of properties holds?

So i'd suggest renaming the operators to reflect the particular properties they check (maybe something like `checkBuiltExpressionIsWellTyped`?)? I think a lot of my confusion in reading this came from trying to understand how this works in the general case because the name is so general, and because the docs here say "BuilderTest implements a framework for PB testing various Builder methods." I think that if it were clear that this only serves to test a single set of properties, it would have been easier for me to grok?

No sure, would be interested to see whether this is just my hangup or if @thpani or @konnov thinks this kind of clarification of the scope would have helped the mas well.

Reply:
> If this helper is only for testing one property, then I'd change the intro to this package to explicitly state this limited scope. It's sounds like it's a utility for writing general purpose property based tests of built expressions, but iiuc from your reply, it just tests that one particular, fixed set of properties holds?

Well, you could invent more convoluted `resultIsExpected` methods, other than the defaults given by `expectEqTyped`. Though for the builder, I'm not exactly sure what kinds of properties you're thinking of.

Reply:
So what I'd think I'd want out of a general purpose PBT utility related to the builder is just a generator or set of generators which, when given types (or combinators expressing types), will generate expressions of that type. But it sounds like this utility is instead for testing something internal to the builder library itself, rather than supporting general purpose PBT using the builder?

It's the intended scope here I'm still not clear on I guess. But it sounds much more specific to me than I'd understand from "BuilderTest implements a framework for PB testing various Builder methods.".

Reply:
> So what I'd think I'd want out of a general purpose PBT utility related to the builder is just a generator or set of generators which, when given types (or combinators expressing types), will generate expressions of that type.

That's trivial, map a `builder.name("x", _)` over a type generator and you get an expression generator for arbitrary types.

> But it sounds like this utility is instead for testing something internal to the builder library itself, rather than supporting general purpose PBT using the builder?

Yes, this is a set of tests _for_ the builder methods, not tests _with_ the builder methods. "[G]eneral purpose PBT using the builder" doesn't really need a library, you just use the builder in tests when you need a `TlaEx`.